File: examples/python/mcp_agent.py
import logging
import os

from langchain.agents import create_agent
from langchain_openai import ChatOpenAI
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from rebuno import Agent
from rebuno.mcp import wrap_mcp_tools

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(name)s %(levelname)s %(message)s"
)
logger = logging.getLogger("mcp")

SYSTEM_PROMPT = (
    "You are a helpful assistant with filesystem access. Use the filesystem tools "
    "to list directories and read files, then answer the user's question."
)

MODEL = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")
MCP_ROOT = os.environ.get("MCP_ROOT", "/tmp")

filesystem = StdioServerParameters(
    command="npx",
    args=["-y", "@modelcontextprotocol/server-filesystem", MCP_ROOT],
)


async def process(query: str) -> dict:
    async with stdio_client(filesystem) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            listed = await session.list_tools()
            tools = wrap_mcp_tools(
                listed.tools,
                call=lambda name, args: session.call_tool(name, args),
                prefix="fs",
            )
            logger.info("MCP tools: %s", [t.__name__ for t in tools])

            llm = ChatOpenAI(model=MODEL, temperature=0)
            graph = create_agent(model=llm, tools=tools, system_prompt=SYSTEM_PROMPT)
            result = await graph.ainvoke({"messages": [{"role": "user", "content": query}]})
            answer = result["messages"][-1].content
            logger.info("done: %s", answer)
            return {"query": query, "answer": answer}


if __name__ == "__main__":
    agent = Agent(
        "mcp",
        secret=os.environ.get("REBUNO_AGENT_SECRET", "mcp-secret"),
        kernel_url=os.environ.get("REBUNO_URL", "http://localhost:8080"),
    )
    logger.info("mcp agent listening on :5003")
    agent.run(process, port=int(os.environ.get("AGENT_PORT", "5003")))

File: examples/typescript/mcp_agent.ts
import { env } from "node:process";

import { createOpenAI } from "@ai-sdk/openai";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { generateText, jsonSchema, stepCountIs, tool, type ToolSet } from "ai";
import { Agent, rebunoFetch, wrapMcpTools, type RebunoTool } from "rebuno";

const SYSTEM_PROMPT =
  "You are a helpful assistant with filesystem access. Use the filesystem tools " +
  "to list directories and read files, then answer the user's question.";

const MODEL = env.OPENAI_MODEL ?? "gpt-4o-mini";
const MCP_ROOT = env.MCP_ROOT ?? "/tmp";

async function process(input: { query: string }): Promise<Record<string, unknown>> {
  const transport = new StdioClientTransport({
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-filesystem", MCP_ROOT],
  });
  const client = new Client({ name: "rebuno-mcp-example", version: "0.1.0" });
  await client.connect(transport);
  try {
    const { tools: listed } = await client.listTools();
    const tools = wrapMcpTools(listed, {
      call: (name, args) => client.callTool({ name, arguments: args }),
      prefix: "fs",
    });
    console.log(`MCP tools: ${tools.map((t) => t.name).join(", ")}`);

    const openai = createOpenAI({ fetch: rebunoFetch });
    const { text } = await generateText({
      model: openai(MODEL),
      system: SYSTEM_PROMPT,
      prompt: input.query,
      tools: asAiTools(tools),
      stopWhen: stepCountIs(10),
    });
    console.log(`done: ${text}`);
    return { query: input.query, answer: text };
  } finally {
    await client.close();
  }
}

/** Expose durable Rebuno tools to the Vercel AI SDK; each execute routes through the kernel. */
function asAiTools(tools: RebunoTool[]): ToolSet {
  return Object.fromEntries(
    tools.map((t) => [
      t.name,
      tool({
        description: t.description,
        inputSchema: jsonSchema((t.inputSchema ?? { type: "object" }) as Record<string, unknown>),
        execute: (args) => t.execute(args as Record<string, unknown>),
      }),
    ]),
  );
}

const agent = new Agent("mcp", {
  secret: env.REBUNO_AGENT_SECRET ?? "mcp-secret",
  baseUrl: env.REBUNO_URL ?? "http://localhost:8080",
});
console.log("mcp agent listening on :5003");
await agent.serve({ port: Number(env.AGENT_PORT ?? 5003) }, process);
